### ISSUE TITLE
Update to Java 10

### ISSUE THREAD

**Issue by DavidWhitlock** [DavidWhitlock/PortlandStateJava#220]
Looks like Java 10 is available.

http://www.oracle.com/technetwork/java/javase/downloads/index.html

I should update to using it and require it for the projects.  I should also ask the CAT to install it on the PSU machines.

**Comment by DavidWhitlock** [comment]
In fact, there are some new library dependencies that require Java 9 or later.

https://stackoverflow.com/questions/44804417/update-javax-json-from-1-0-2-to-1-1

**Comment by DavidWhitlock** [comment]
The game's afoot!

https://maarten.mulders.tk/blog/2018/01/migrating-to-java9.html

https://blog.codefx.org/tools/maven-on-java-9/

**Comment by DavidWhitlock** [comment]
Looks like I'll need to remove cobertura because it doesn't support Java 9 (which no lonegr has tools.jar).

According to [this](https://en.wikipedia.org/wiki/Java_Code_Coverage_Tools), JaCoCo supports Java 9 and 10.

**Comment by DavidWhitlock** [comment]
It looks like GWT 2.8.2 supports Java 9: http://www.gwtproject.org/release-notes.html#Release_Notes_2_8_2

I should probably upgrade from 2.7.0 in a separate branch off of `master`

**Comment by DavidWhitlock** [comment]
Now that I've gotten rid of cobertura and upgraded GWT, I'm seeing this error while running the surefire Maven plugin in the web examples:

```
java.lang.Error: Circular loading of URL stream handler providers detected
        at java.base/java.net.URL.lookupViaProviders(URL.java:1299)
        at java.base/java.net.URL.getURLStreamHandler(URL.java:1377)
        at java.base/java.net.URL.<init>(URL.java:633)
        at java.base/java.net.URL.<init>(URL.java:523)
        at java.base/jdk.internal.loader.URLClassPath$JarLoader.parseClassPath(URLClassPath.java:1026)
        at java.base/jdk.internal.loader.URLClassPath$JarLoader.getClassPath(URLClassPath.java:1006)
        at java.base/jdk.internal.loader.URLClassPath.getLoader(URLClassPath.java:380)
        at java.base/jdk.internal.loader.URLClassPath.access$000(URLClassPath.java:84)
        at java.base/jdk.internal.loader.URLClassPath$1.next(URLClassPath.java:270)
        at java.base/jdk.internal.loader.URLClassPath$1.hasMoreElements(URLClassPath.java:281)
        at java.base/jdk.internal.loader.BuiltinClassLoader$1.hasNext(BuiltinClassLoader.java:356)
        at java.base/jdk.internal.loader.BuiltinClassLoader$1.hasMoreElements(BuiltinClassLoader.java:364)
        at java.base/java.lang.CompoundEnumeration.next(ClassLoader.java:2921)
        at java.base/java.lang.CompoundEnumeration.hasMoreElements(ClassLoader.java:2930)
        at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1197)
        at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1215)
        at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1259)
        at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1294)
        at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1379)
        at java.base/java.net.URL$1.getNext(URL.java:1266)
        at java.base/java.net.URL$1.hasNext(URL.java:1281)
        at java.base/java.net.URL$2.run(URL.java:1307)
        at java.base/java.net.URL$2.run(URL.java:1304)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.base/java.net.URL.lookupViaProviders(URL.java:1303)
        at java.base/java.net.URL.getURLStreamHandler(URL.java:1377)
        at java.base/java.net.URL.<init>(URL.java:633)
        at java.base/java.net.URL.<init>(URL.java:523)
        at java.base/jdk.internal.loader.URLClassPath$JarLoader.parseClassPath(URLClassPath.java:1026)
        at java.base/jdk.internal.loader.URLClassPath$JarLoader.getClassPath(URLClassPath.java:1006)
        at java.base/jdk.internal.loader.URLClassPath.getLoader(URLClassPath.java:380)
        at java.base/jdk.internal.loader.URLClassPath.getResource(URLClassPath.java:243)
        at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:696)
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:622)
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:580)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:185)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:496)
        at org.apache.maven.surefire.util.ReflectionUtils.loadClass(ReflectionUtils.java:241)
        at org.apache.maven.surefire.util.ReflectionUtils.instantiateOneArg(ReflectionUtils.java:133)
        at org.apache.maven.surefire.booter.ForkedBooter.createProviderInCurrentClassloader(ForkedBooter.java:398)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:378)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:340)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:413)
```

Oddly enough, it doesn't reproduce on Travis.  I'm trying to figure out why.

**Comment by DavidWhitlock** [comment]
I should also update the javadoc to use the new HTML 5 features, including search.

**Comment by DavidWhitlock** [comment]
Ooh.  Also, the JShell Console...which is inside IntelliJ:

![image](https://user-images.githubusercontent.com/2935912/40403029-f1d54192-5e03-11e8-8026-35a3d3a8d149.png)

**Comment by DavidWhitlock** [comment]
Some videos that I've watched to learn about Java 9 and 10:
   * [Real World Java 9](https://www.youtube.com/watch?v=QqmQ_0tV978)
   * [Java 10 and IntelliJ IDEA](https://www.youtube.com/watch?v=H-LzZofU3zk)
   * [Exploring Java 9: The Key Parts by Venkat Subramaniam](https://www.youtube.com/watch?v=Yacu1yUktjY)

**Comment by DavidWhitlock** [comment]
I wonder if the `Circular loading of URL stream handler providers detected` problem has something to do with the shade Maven plugin that creates an uber jar.

**Comment by DavidWhitlock** [comment]
Actually, that "Circular loading" problem appeared to be caused by the fact that the family tree application's jar file had a full http URL in it.  Weird.  See https://github.com/DavidWhitlock/PortlandStateJava/commit/63981c3a8d01755cb85f9b980940ad7fb43252ea

**Comment by DavidWhitlock** [comment]
It looks like I might have to migrate from findbugs to spotbugs (https://spotbugs.github.io/spotbugs-maven-plugin) in order to get Java 9/10 support.

**Comment by DavidWhitlock** [comment]
Looks like I've got it building with Java 9.  Now on to Java 10.

I ran into this while running spotbugs: https://github.com/spotbugs/spotbugs/issues/593

**Comment by DavidWhitlock** [comment]
Also hit this defect in the javadoc Plugin: https://issues.apache.org/jira/browse/MJAVADOC-517

**Comment by DavidWhitlock** [comment]
I should probably run javadoc with -html5 to get the newer API documentation format.

**Comment by DavidWhitlock** [comment]
Changes for this issue have been merged into `master` and have been deployed to Bintray.  Now I just need to make sure that the archetypes still work.  And I need to ask the PSU CAT to install Java 10 on the Linux machines.

**Comment by DavidWhitlock** [comment]
I still need to get the javadocs generating.  So, I won't call this all the way done yet.

**Comment by DavidWhitlock** [comment]
I got the site to generate, but now I'm running into https://github.com/trajano/wagon-git/issues/22 with the trajano wagon-git plugin.  I'll see if I can update some dependencies and get it to work.

**Comment by DavidWhitlock** [comment]
Actually, I was able to get deploying to GitHub Pages to work by following the configuration in the parent POM of spotbugs (which deploys its site to GitHub).

`com/github/hazendaz/base-parent/17/base-parent-17.pom`

**Comment by DavidWhitlock** [comment]
I ran into this issue when I instantiated the phonebill web archetype and ran `mvn jetty:run`

https://github.com/eclipse/jetty.project/issues/2397

Oh, but it says that the version of Jetty that I'm using is `jetty-9.2.10.v20150310` which is super-old.  That would explain when the integration tests worked fine.  They used Jetty 9.4.8.v20171121.  

It looks like 9.4.11.v20180605 is available.  I wonder why I hadn't updated sooner.
